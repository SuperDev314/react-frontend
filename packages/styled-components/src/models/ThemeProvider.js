// @flow
import React, { useContext, type Element, type Context } from 'react';
import StyledError from '../utils/error';
import isFunction from '../utils/isFunction';

export type Theme = { [key: string]: mixed };

type Props = {
  children?: Element<any>,
  theme: Theme | ((outerTheme: Theme) => void),
};

export const ThemeContext: Context<Theme | void> = React.createContext();

export const ThemeConsumer = ThemeContext.Consumer;

function useMergedTheme(theme: Theme | ((outerTheme?: Theme) => Theme), outerTheme?: Theme): Theme {
  if (isFunction(theme)) {
    const mergedTheme = theme(outerTheme);

    if (
      process.env.NODE_ENV !== 'production' &&
      (mergedTheme === null || Array.isArray(mergedTheme) || typeof mergedTheme !== 'object')
    ) {
      throw new StyledError(7);
    }

    return mergedTheme;
  }

  if (theme === null || Array.isArray(theme) || typeof theme !== 'object') {
    throw new StyledError(8);
  }

  return outerTheme ? { ...outerTheme, ...theme } : theme;
}

/**
 * Provide a theme to an entire react component tree via context
 */
export default function ThemeProvider(props: Props) {
  const outerTheme = useContext(ThemeContext);
  // NOTE: can't really memoize with props.theme as that'd cause incorrect memoization when it's a function
  const themeContext = useMergedTheme(props.theme, outerTheme);

  if (!props.children) {
    return null;
  }

  return (
    <ThemeContext.Provider value={themeContext}>
      {React.Children.only(props.children)}
    </ThemeContext.Provider>
  );
}
