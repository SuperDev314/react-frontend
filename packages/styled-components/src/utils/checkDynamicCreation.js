// @flow

import { useRef } from 'react';

const invalidHookCallRe = /invalid hook call/i;
const seen = new Set();

export const checkDynamicCreation = (displayName: string, componentId?: string) => {
  if (process.env.NODE_ENV !== 'production') {
    const parsedIdString = componentId ? ` with the id of "${componentId}"` : '';
    const message =
      `The component ${displayName}${parsedIdString} has been created dynamically.\n` +
      "You may see this warning because you've called styled inside another component.\n" +
      'To resolve this only create new StyledComponents outside of any render method and function component.';

    try {
      // We purposefully call `useRef` outside of a component and expect it to throw
      // If it doesn't, then we're inside another component.
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useRef();

      if (!seen.has(message)) {
        // eslint-disable-next-line no-console
        console.warn(message);
        seen.add(message);
      }
    } catch (error) {
      // The error here is expected, since we're expecting anything that uses `checkDynamicCreation` to
      // be called outside of a React component.
      if (invalidHookCallRe.test(error.message)) {
        // This shouldn't happen, but resets `warningSeen` if we had this error happen intermittently
        seen.delete(message);
      }
    }
  }
};
