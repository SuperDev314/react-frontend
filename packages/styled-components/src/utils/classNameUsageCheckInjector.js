// @flow
import ReactDOM from 'react-dom';
import getComponentName from './getComponentName';

const didWarnAboutClassNameUsage = new Set();

export default (target: Object) => {
  let elementClassName = '';

  const targetCDM = target.componentDidMount;

  // eslint-disable-next-line no-param-reassign
  target.componentDidMount = function componentDidMount() {
    if (typeof targetCDM === 'function') {
      targetCDM.call(this);
    }

    const forwardTarget = this.props.forwardedComponent.target;

    if (
      (target.props && target.props.suppressClassNameWarning) ||
      (target.attrs && target.attrs.suppressClassNameWarning) ||
      didWarnAboutClassNameUsage.has(forwardTarget)
    ) {
      return;
    }

    didWarnAboutClassNameUsage.add(forwardTarget);

    const classNames = elementClassName
      .replace(/ +/g, ' ')
      .trim()
      .split(' ');
    // eslint-disable-next-line react/no-find-dom-node
    const node: Element | null = (ReactDOM.findDOMNode(this): any);
    const selector = classNames.map(s => `.${s}`).join('');

    if (
      node &&
      node.nodeType === 1 &&
      !classNames.every(className => node.classList && node.classList.contains(className)) &&
      !node.querySelector(selector)
    ) {
      console.warn(
        `It looks like you've wrapped styled() around your React component (${getComponentName(
          forwardTarget
        )}), but the className prop is not being passed down to a child. No styles will be rendered unless className is composed within your React component.`
      );
    }
  };

  const prevRenderInner = target.renderInner;

  // eslint-disable-next-line no-param-reassign
  target.renderInner = function renderInner(...args) {
    const element = prevRenderInner.apply(this, args);

    elementClassName = element.props.className;

    return element;
  };
};
