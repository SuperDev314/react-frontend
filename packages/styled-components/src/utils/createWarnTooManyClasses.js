// @flow

export const LIMIT = 50;

export default (displayName: string) => {
  let generatedClasses = {};
  let warningSeen = false;

  return (className: string) => {
    if (!warningSeen) {
      generatedClasses[className] = true;
      if (Object.keys(generatedClasses).length >= LIMIT) {
        // Unable to find latestRule in test environment.
        /* eslint-disable no-console, prefer-template */
        console.warn(
          `Over ${LIMIT} classes were generated for component ${displayName}. This happens when some of the props you use for styling have many potential values and we need to make a new CSS class for each variant. Over time the stylesheet will grow and slow down your app.\n` +
            'For these particular CSS rules with high dynamicity, consider using the attrs() method together with a style object.\n' +
            'Example:\n' +
            '  const Component = styled.div.attrs(props => ({\n' +
            '    style: {\n' +
            '      background: props.background,\n' +
            '    },\n' +
            '  }))`width: 100%;`\n\n' +
            '  <Component />'
        );
        warningSeen = true;
        generatedClasses = {};
      }
    }
  };
};
