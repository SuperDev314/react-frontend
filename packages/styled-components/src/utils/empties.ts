export const EMPTY_ARRAY = (Object.freeze([]) as unknown) as Readonly<any[]>;
export const EMPTY_OBJECT = Object.freeze({});
