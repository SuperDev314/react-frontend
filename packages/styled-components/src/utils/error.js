// @flow
import errorMap from './errors';

const ERRORS = process.env.NODE_ENV !== 'production' ? errorMap : {};

/**
 * super basic version of sprintf
 */
function format(...args) {
  let a = args[0];
  const b = [];

  for (let c = 1, len = args.length; c < len; c += 1) {
    b.push(args[c]);
  }

  b.forEach(d => {
    a = a.replace(/%[a-z]/, d);
  });

  return a;
}

/**
 * Create an error file out of errors.md for development and a simple web link to the full errors
 * in production mode.
 */
export default class StyledComponentsError extends Error {
  constructor(code: string | number, ...interpolations: Array<any>) {
    if (process.env.NODE_ENV === 'production') {
      super(
        `An error occurred. See https://github.com/styled-components/styled-components/blob/master/packages/styled-components/src/utils/errors.md#${code} for more information.${
          interpolations.length > 0 ? ` Additional arguments: ${interpolations.join(', ')}` : ''
        }`
      );
    } else {
      super(format(ERRORS[code], ...interpolations).trim());
    }
  }
}
