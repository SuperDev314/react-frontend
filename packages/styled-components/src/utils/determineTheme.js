// @flow
import { EMPTY_OBJECT } from './empties';

type Props = {
  theme?: any,
};

export default (props: Props, providedTheme: any, defaultProps: any = EMPTY_OBJECT) => {
  return (props.theme !== defaultProps.theme && props.theme) || providedTheme || defaultProps.theme;
};
