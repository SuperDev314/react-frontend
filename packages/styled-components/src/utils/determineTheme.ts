import { ExecutionProps } from '../types';
import { EMPTY_OBJECT } from './empties';

export default function determineTheme(
  props: ExecutionProps,
  providedTheme: any,
  defaultProps: any = EMPTY_OBJECT
) {
  return (props.theme && props.theme !== defaultProps.theme) || providedTheme || defaultProps.theme;
}
