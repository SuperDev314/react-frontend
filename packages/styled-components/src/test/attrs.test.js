// @flow
import React, { Component } from 'react';
import TestRenderer from 'react-test-renderer';
import ThemeProvider from '../models/ThemeProvider';

import { resetStyled, expectCSSMatches } from './utils';

let styled;

describe('attrs', () => {
  beforeEach(() => {
    jest.spyOn(console, 'warn');
    styled = resetStyled();
  });

  it('work fine with an empty object', () => {
    const Comp = styled.div.attrs({})``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
  });

  it('pass a simple attr', () => {
    const Comp = styled.button.attrs({
      type: 'button',
    })``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
  });

  it('pass a React component', () => {
    // $FlowFixMe
    class ReactComponent extends Component {
      render() {
        return <p>React Component</p>;
      }
    }

    const Button = ({ component: ChildComponent }) => (
      <button>
        <ChildComponent />
      </button>
    );

    const Comp = styled(Button).attrs({
      component: ReactComponent,
    })``;

    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
  });

  it('call an attr function', () => {
    console.warn.mockImplementation(() => {});
    const Comp = styled.button.attrs({
      as: () => 'div',
    })``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();

    // deprecation message for fn as attrs object key
    expect(console.warn).toHaveBeenCalled();
  });

  it('pass a fn to attrs', () => {
    const Comp = styled.button.attrs(props => ({
      as: props.renderAs,
    }))``;
    expect(TestRenderer.create(<Comp renderAs="div" />).toJSON()).toMatchSnapshot();
  });

  it('function form allows access to theme', () => {
    const Comp = styled.button.attrs(props => ({
      'data-color': props.theme.color,
    }))``;

    expect(
      TestRenderer.create(
        <ThemeProvider theme={{ color: 'red' }}>
          <Comp />
        </ThemeProvider>
      ).toJSON()
    ).toMatchInlineSnapshot(`
<button
  className="sc-a b"
  data-color="red"
/>
`);
  });

  it('defaultProps are merged into what function attrs receives', () => {
    const Comp = styled.button.attrs(props => ({
      'data-color': props.theme.color,
    }))``;

    Comp.defaultProps = {
      theme: {
        color: 'red',
      },
    };

    expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
<button
  className="sc-a b"
  data-color="red"
/>
`);
  });

  it('pass props to the attr function', () => {
    const Comp = styled.button.attrs({
      type: props => (props.submit ? 'submit' : 'button'),
    })``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
    expect(TestRenderer.create(<Comp submit />).toJSON()).toMatchSnapshot();
  });

  it('should replace attrs with props', () => {
    const Comp = styled.button.attrs({
      type: props => (props.submit ? 'submit' : 'button'),
      tabIndex: 0,
    })``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
    expect(TestRenderer.create(<Comp type="reset" />).toJSON()).toMatchSnapshot();
    expect(TestRenderer.create(<Comp type="reset" tabIndex="-1" />).toJSON()).toMatchSnapshot();
  });

  it('should merge className', () => {
    const Comp = styled.div.attrs({
      className: 'meow nya',
    })``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
  });

  it('should merge className even if its a function', () => {
    const Comp = styled.div.attrs({
      className: props => `meow ${props.purr ? 'purr' : 'nya'}`,
    })``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
    expect(TestRenderer.create(<Comp purr />).toJSON()).toMatchSnapshot();
  });

  it('should work with data and aria attributes', () => {
    const Comp = styled.div.attrs({
      'data-foo': 'bar',
      'aria-label': 'A simple FooBar',
    })``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
  });

  it('merge attrs', () => {
    const Comp = styled.button
      .attrs({
        type: 'button',
        tabIndex: 0,
      })
      .attrs({
        type: 'submit',
      })``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
  });

  it('merge fn attrs', () => {
    const Comp = styled.button
      .attrs(() => ({
        type: 'button',
        tabIndex: 0,
      }))
      .attrs(() => ({
        type: 'submit',
      }))``;

    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
  });

  it('merge attrs when inheriting SC', () => {
    const Parent = styled.button.attrs({
      type: 'button',
      tabIndex: 0,
    })``;
    const Child = styled(Parent).attrs({
      type: 'submit',
    })``;
    expect(TestRenderer.create(<Child />).toJSON()).toMatchSnapshot();
  });

  it('pass attrs to style block', () => {
    /* Would be a React Router Link in real life */
    const Comp = styled.a.attrs({
      href: '#',
      'data-active-class-name': '--is-active',
    })`
      color: blue;
      &.${props => props['data-active-class-name']} {
        color: red;
      }
    `;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
    expectCSSMatches('.b { color:blue; } .b.--is-active { color:red; }');
  });

  it('should pass through children as a normal prop', () => {
    const Comp = styled.div.attrs({
      children: 'Probably a bad idea',
    })``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
  });

  it('should pass through complex children as well', () => {
    const Comp = styled.div.attrs({
      children: <span>Probably a bad idea</span>,
    })``;
    expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
  });

  it('should override children of course', () => {
    const Comp = styled.div.attrs({
      children: <span>Amazing</span>,
    })``;
    expect(TestRenderer.create(<Comp>Something else</Comp>).toJSON()).toMatchSnapshot();
  });

  it('should shallow merge "style" prop + attr instead of overwriting', () => {
    const Paragraph = styled.p.attrs({
      style: props => ({
        ...props.style,
        fontSize: `${props.fontScale}em`,
      }),
    })``;

    class Text extends React.Component {
      state = {
        // Assume that will be changed automatically
        // according to the dimensions of the container
        fontScale: 4,
      };

      render() {
        return (
          <Paragraph
            className={this.props.className}
            style={this.props.style}
            fontScale={this.state.fontScale}
          >
            {this.props.children}
          </Paragraph>
        );
      }
    }

    const BlueText = styled(Text).attrs({
      style: () => ({
        color: 'blue',
      }),
    })``;

    expect(TestRenderer.create(<BlueText>Hello</BlueText>).toJSON()).toMatchSnapshot();
  });

  it('does not pass non html tags to HTML element', () => {
    const Comp = styled.div`
      color: ${props => props.textColor};
    `;

    const StyledComp = styled(Comp).attrs({
      textColor: 'red',
    })``;
    expect(TestRenderer.create(<StyledComp />).toJSON()).toMatchSnapshot();
  });

  describe('warnings', () => {
    beforeEach(() => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});
    });

    it('warns upon use of a Stateless Functional Component as a prop', () => {
      const Inner = () => <div />;
      const Comp = styled.div.attrs({ component: Inner })``;

      TestRenderer.create(<Comp />);

      expect(console.warn.mock.calls[1][0]).toMatchInlineSnapshot(`
"It looks like you've used a non styled-component as the value for the \\"component\\" prop in an object-form attrs constructor of \\"styled.div\\".
You should use the new function-form attrs constructor which avoids this issue: attrs(props => ({ yourStuff }))
To continue using the deprecated object syntax, you'll need to wrap your component prop in a function to make it available inside the styled component (you'll still get the deprecation warning though.)
For example, { component: () => InnerComponent } instead of { component: InnerComponent }"
`);
    });

    it('warns for using fns as attrs object keys', () => {
      const Comp = styled.div.attrs({ 'data-text-color': props => props.textColor })``;

      TestRenderer.create(<Comp textColor="blue" />);

      expect(console.warn.mock.calls[0][0]).toMatchInlineSnapshot(
        `"Functions as object-form attrs({}) keys are now deprecated and will be removed in a future version of styled-components. Switch to the new attrs(props => ({})) syntax instead for easier and more powerful composition. The attrs key in question is \\"data-text-color\\" on component \\"styled.div\\"."`
      );
    });
  });
});
