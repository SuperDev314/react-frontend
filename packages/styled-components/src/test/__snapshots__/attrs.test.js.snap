// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attrs call an attr function 1`] = `
<div
  className="sc-a b"
/>
`;

exports[`attrs does not pass non html tags to HTML element 1`] = `
<div
  className="sc-a sc-b c"
/>
`;

exports[`attrs merge attrs 1`] = `
<button
  className="sc-a b"
  tabIndex={0}
  type="submit"
/>
`;

exports[`attrs merge attrs when inheriting SC 1`] = `
<button
  className="sc-a sc-b c"
  tabIndex={0}
  type="submit"
/>
`;

exports[`attrs merge fn attrs 1`] = `
<button
  className="sc-a b"
  tabIndex={0}
  type="submit"
/>
`;

exports[`attrs pass a React component 1`] = `
<button>
  <p>
    React Component
  </p>
</button>
`;

exports[`attrs pass a fn to attrs 1`] = `
<div
  className="sc-a b"
/>
`;

exports[`attrs pass a simple attr 1`] = `
<button
  className="sc-a b"
  type="button"
/>
`;

exports[`attrs pass attrs to style block 1`] = `
<a
  className="sc-a b"
  data-active-class-name="--is-active"
  href="#"
/>
`;

exports[`attrs pass props to the attr function 1`] = `
<button
  className="sc-a b"
  type="button"
/>
`;

exports[`attrs pass props to the attr function 2`] = `
<button
  className="sc-a b"
  type="submit"
/>
`;

exports[`attrs should merge className 1`] = `
<div
  className="sc-a meow nya b"
/>
`;

exports[`attrs should merge className even if its a function 1`] = `
<div
  className="sc-a meow nya b"
/>
`;

exports[`attrs should merge className even if its a function 2`] = `
<div
  className="sc-a meow purr b"
/>
`;

exports[`attrs should override children of course 1`] = `
<div
  className="sc-a b"
>
  Something else
</div>
`;

exports[`attrs should pass through children as a normal prop 1`] = `
<div
  className="sc-a b"
>
  Probably a bad idea
</div>
`;

exports[`attrs should pass through complex children as well 1`] = `
<div
  className="sc-a b"
>
  <span>
    Probably a bad idea
  </span>
</div>
`;

exports[`attrs should replace attrs with props 1`] = `
<button
  className="sc-a b"
  tabIndex={0}
  type="button"
/>
`;

exports[`attrs should replace attrs with props 2`] = `
<button
  className="sc-a b"
  tabIndex={0}
  type="reset"
/>
`;

exports[`attrs should replace attrs with props 3`] = `
<button
  className="sc-a b"
  tabIndex="-1"
  type="reset"
/>
`;

exports[`attrs should shallow merge "style" prop + attr instead of overwriting 1`] = `
<p
  className="sc-b c sc-a d"
  style={
    Object {
      "color": "blue",
      "fontSize": "4em",
    }
  }
>
  Hello
</p>
`;

exports[`attrs should work with data and aria attributes 1`] = `
<div
  aria-label="A simple FooBar"
  className="sc-a b"
  data-foo="bar"
/>
`;

exports[`attrs work fine with an empty object 1`] = `
<div
  className="sc-a b"
/>
`;
